jax/abstract_arrays.py

```diff
@@ -154,7 +160,8 @@ def __init__(self, val, weak_type=False):
                                         weak_type=weak_type)
     # Note: canonicalized self.dtype doesn't necessarily match self.val
     self.val = val
-    assert self.dtype != onp.dtype('O')
+
+    assert self.dtype != onp.dtype('O') or is_polymorphic(val)
```

Comment:
I'd rather avoid storing object-dtype arrays, even if we create them as intermediates (by making `onp` operate on tuples of our Poly instances).

That said, I don't see why we'd need to create object-dtype arrays of _values_, just of shapes. So why would this check be needed?

Reply:
Also, the definition of `is_polymorphic` looks like it's meant to apply to shapes, not values, but here it's being applied to a value. I'm confused!

Reply:
As I understand it, the argument for `gather`'s `start_indices` can an abstract value, i. e. when  jitting dynamic indexing. On the other hand, it can be a polymorphic value, i. e. `'n-1'` when indexing `x[-1]` where `x` is of polymorphic shape `'n'`. [This line](https://github.com/google/jax/pull/2017/files#diff-08adae3ce53df47932e06971edb68642R3030) reveals more of what's going on. This is the case where we need this atm AFAIK. Let me know if you have an idea how to avoid this.

Reply:
The abstract value in that case should be a ShapedArray of dtype int, right? A ConcreteArray represents a specific array value, but we have no concrete value here.

I think we need to add abstraction handlers for Poly instances, which will produce something at the Shaped level unless they represent constant polynomials. We don't want to treat them as concrete arrays of dtype object; that'd be mixing in a new way of representing sets of possible values, but we already have a way to do that.

Reply:
@mattjj That makes sense. I think this would require to first solve [the problem we discussed earlier](https://github.com/google/jax/pull/2017#issue-363890835) though: `idx` is passed into `index_to_gather` without being wrapped into a trace for cases like `x[-1]`, resulting in the polymorphic value `n-1` for `i` [here](https://github.com/google/jax/pull/2017/files#diff-a4404c86bdd5bdfd0128f72b7aa1d40bR2965). Representing `i` as `ShapedArray((), int32)` would therefore currently break further processing due to abstract values getting passed into implementations. Lifting `i` back into a trace should allow getting rid of the current workaround.

Reply:
**Update:** I found a workaround for `gather` that is much simpler and doesn't require trace lifting. It allowed me to retighten the assertion above to the state before and remove other special cases, further minimizing the diff.